x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -383,7 +387,7 @@ static class AuditableRestRequest extends AuditableRequest {
 
         @Override
         void authenticationSuccess(Authentication authentication) {
-            auditTrail.authenticationSuccess(requestId, authentication, request);
+            // REST requests are audited in the {@code SecurityRestFilter} because they need access to the request body
```

Comment:
This feels off ... it seems that we should have a reference to the AuditableRestRequest when we want to audit... even if we need to check instance type and downcast I think would be preferable to ignoring our own contract.

Reply:
I agree this is not great, but I struggle to find better options...

Fundamentally, the `AuditableRequest#authenticationSuccess` is called after authentication was successful, and we don't want to audit HTTP requests that way (because we do the auditing in the `SecurityRestFilter#handleRequest`, for the `RestRequest`).

It is possible to alternatively have `auditTrail.authenticationSuccess(requestId, authentication, request)` be a noop. Or, to not have the `context.getRequest().authenticationSuccess(authentication)` called in the `AuthenticatorChain` if the request is a HTTP request. I think these options are more hackish.

What do you think?
We could also file this objection of yours, and come back to it later when the puzzle pieces are in place?

Reply:
> Wrap RestRequest as AuditableRestRequest earlier in SecurityRestFilter so that auditing can be consistently done with request.authenticationSuccess.

I think this is glancing over important details.
Where are you proposing we invoke `AuditableRequest#authenticationSuccess`?
Currently, the callsite is in the `AuthenticationService` just after authn completed successfully AND has just been written to the thread context. The proposed implementation doesn't change the call site, but for REST requests only, turns it into a noop. It does that because the REST request body contents might be needed for auditing authn success and that won't be available, at the current callsite (i.e. after authN just finished and its result been written to the threadcontext). Instead, the proposed implementation, creates a new method in the `AuditTrail` class specifically for dealing with auditing of REST requests (to cover for the fact that now `AuditableRequest#authenticationSuccess` is a noop for REST requests) and calls that in the `SecurityRestFilter`.

I don't see an easy way to move the callsite of `AuditableRequest#authenticationSuccess`, which is what I think you're suggesting, because the `AuditableRequest` is tied to the `AuthenticationService` and it's lifetime doesn't go beyond the authN flow.
So maybe an option is to remove the `AuditableRequest#authenticationSuccess` completely and find a place to audit authN success directly via the `AuditTrail` class for both REST and transport requests?

Reply:
@albertzaharovits Thanks for summarizing our discussion. I think it might be useful to somehow make the summary available in the code, e.g. maybe add a comment with link to your above comment.

Reply:
Pushed  8f5e1b337f37850c .